share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -17,58 +17,536 @@
+     * <p>
+     * This method as called by the ShareCoordinatorService will be provided with
+     * the request data which covers only key i.e. group1:topic1:partition1. The implementation
+     * below was done keeping this in mind.
+     *
+     * @param request - WriteShareGroupStateRequestData for a single key
+     * @param offset  - offset to read from the __share_group_state topic partition
+     * @return CoordinatorResult(records, response)
+     */
     public ReadShareGroupStateResponseData readState(ReadShareGroupStateRequestData request, Long offset) {
-        throw new RuntimeException("Not implemented");
+        log.debug("Read request dump - {}", request);
+        // records to read (with the key of snapshot type), response to caller
+        // only one key will be there in the request by design
+        Optional<ReadShareGroupStateResponseData> error = maybeGetReadStateError(request, offset);
+        if (error.isPresent()) {
+            return error.get();
+        }
+
+        Uuid topicId = request.topics().get(0).topicId();
+        int partition = request.topics().get(0).partitions().get(0).partition();
+        int leaderEpoch = request.topics().get(0).partitions().get(0).leaderEpoch();
+
+        SharePartitionKey coordinatorKey = SharePartitionKey.getInstance(request.groupId(), topicId, partition);
+
+        if (!shareStateMap.containsKey(coordinatorKey)) {
+            return ReadShareGroupStateResponse.toResponseData(
+                topicId,
+                partition,
+                PartitionFactory.DEFAULT_START_OFFSET,
+                PartitionFactory.DEFAULT_STATE_EPOCH,
+                Collections.emptyList()
+            );
+        }
+
+        ShareGroupOffset offsetValue = shareStateMap.get(coordinatorKey, offset);
+
+        if (offsetValue == null) {
+            // Returning an error response as the snapshot value was not found
+            return ReadShareGroupStateResponse.toErrorResponseData(
+                topicId,
+                partition,
+                Errors.UNKNOWN_SERVER_ERROR,
+                "Data not found for topic {}, partition {} for group {}, in the in-memory state of share coordinator"
+            );
+        }
+
+        List<ReadShareGroupStateResponseData.StateBatch> stateBatches = (offsetValue.stateBatches() != null && !offsetValue.stateBatches().isEmpty()) ?
+            offsetValue.stateBatches().stream().map(
+                stateBatch -> new ReadShareGroupStateResponseData.StateBatch()
+                    .setFirstOffset(stateBatch.firstOffset())
+                    .setLastOffset(stateBatch.lastOffset())
+                    .setDeliveryState(stateBatch.deliveryState())
+                    .setDeliveryCount(stateBatch.deliveryCount())
+            ).collect(java.util.stream.Collectors.toList()) : Collections.emptyList();
+
+        // Updating the leader map with the new leader epoch
+        leaderEpochMap.put(coordinatorKey, leaderEpoch);
```

Comment:
Why are we updating the in-memory state during a read? Is this part of the leader epoch fencing?

Reply:
Yes it is.

Caller might issue a read without ever making a write request.

Reply:
Will this logic be replaced by the init RPC once it is added?

In general, we should not be updating our log based in-memory state on reads. It's probably safe in this case just based on the nature of leader epoch, but it is definitely atypical.

Reply:
No, this will stay.
As per the spec initialize RPC does not include a leader epoch value which we can reference
https://cwiki.apache.org/confluence/display/KAFKA/KIP-932%3A+Queues+for+Kafka#KIP932:QueuesforKafka-InitializeShareGroupStateAPI